test/e2e_node/garbage_collector_test.go

```diff
@@ -0,0 +1,153 @@
+http://kubernetes.io/docs/admin/garbage-collection/
+TODO (dashpole): Once the Container Runtime Interface (CRI) is complete, generalize to other runtimes (other than docker)
+TODO (dashpole): Once dynamic config is possible, test different values for maxPerPodContainer and maxContainers
+*/
+const (
+	defaultDockerEndpoint                = "unix:///var/run/docker.sock"
+	defaultRuntimeRequestTimeoutDuration = 2 * time.Minute
+	maxPerPodContainer                   = 2
+	maxTotalContainers                   = -1
+
+	runtimePollInterval = 10 * time.Second
+	deleteTimeout       = 2 * time.Minute
+
+	containerNamePrefix = "gc-test-container"
+)
+
+var _ = framework.KubeDescribe("GarbageCollect [Serial]", func() {
+	f := framework.NewDefaultFramework("garbage-collect-test")
+	runtime := docker.ConnectToDockerOrDie(defaultDockerEndpoint, defaultRuntimeRequestTimeoutDuration)
+
+	Context("", func() {
+		AfterEach(func() {
+			glog.Infof("Summary of node events during the garbage collection test:")
+			err := framework.ListNamespaceEvents(f.ClientSet, f.Namespace.Name)
+			framework.ExpectNoError(err)
+			glog.Infof("Summary of pod events during the garbage collection test:")
+			err = framework.ListNamespaceEvents(f.ClientSet, "")
+			framework.ExpectNoError(err)
+		})
+		maxPerPodContainerGCTest(f, runtime, 6, 1)
+		maxPerPodContainerGCTest(f, runtime, 2, 4)
+		maxPerPodContainerGCTest(f, runtime, 3, 6)
+		maxPerPodContainerGCTest(f, runtime, 4, 1)
+	})
+
+})
+
+// Tests that maxPerPodContainer limits the number of non-garbage collected containers for each container created.
+func maxPerPodContainerGCTest(f *framework.Framework, runtime docker.DockerInterface, restartCount int, numContainers int) {
+	Context(fmt.Sprintf("when we create and delete pods with containers %v times", restartCount), func() {
+		podName := fmt.Sprintf("gc-test-pod%d.%d", numContainers, restartCount)
+		BeforeEach(func() {
+			f.PodClient().CreateSync(getPodWithRestartCount(podName, numContainers, restartCount))
+			time.Sleep(10 * time.Second)
+		})
+		It(fmt.Sprintf("should eventually garbage collect containers that have been recreated more than %v times", maxPerPodContainer), func() {
+			Eventually(func() error {
+				dockerContainers, err := getDockerContainers(runtime)
+				glog.Infof("checking to see if containers have been garbage collected... We currently have %v containers", len(dockerContainers))
+				if err != nil {
+					return err
+				}
+				if maxTotalContainers > 0 && len(dockerContainers) > maxTotalContainers {
+					return fmt.Errorf("expected total number of containers: %v, to be <= maxTotalContainers: %v", len(dockerContainers), maxTotalContainers)
+				}
+				if len(dockerContainers) > maxPerPodContainer*numContainers {
+					return fmt.Errorf("expected total number of containers: %v, to be <= maxPerPodContainer * numContainers %v", len(dockerContainers), maxPerPodContainer*numContainers)
```

Comment:
Some test we may want to add in this PR:
1) We should create multiple pods, and make sure that containers of each pod don't exceed `maxPerPodContainer`.
2) We should make sure that at least one historical container will be kept for each container.
3) We should make sure that after pod is deleted, all containers should be garbaged collected eventually.